graphbolt/src/csc_sampling_graph.cc

```diff
@@ -131,6 +131,117 @@ c10::intrusive_ptr<SampledSubgraph> CSCSamplingGraph::InSubgraph(
           : torch::nullopt);
 }
 
+/**
+ * @brief Calculate the number of the neighbors to be picked for the given node.
+ *
+ * @param fanout The number of edges to be sampled for each node. It should be
+ * >= 0 or -1.
+ *  - When the value is -1, all neighbors will be chosen for sampling. It is
+ * equivalent to selecting all neighbors with non-zero probability when the
+ * fanout is >= the number of neighbors (and replacement is set to false).
+ *  - When the value is a non-negative integer, it serves as a minimum
+ * threshold for selecting neighbors.
+ * @param replace Boolean indicating whether the sample is performed with or
+ * without replacement. If True, a value can be selected multiple times.
+ * Otherwise, each value can be selected only once.
+ * @param probs_or_mask Optional tensor containing the (unnormalized)
+ * probabilities associated with each neighboring edge of a node in the original
+ * graph. It must be a 1D floating-point tensor with the number of elements
+ * equal to the number of edges in the graph.
+ * @param offset The starting edge ID for the connected neighbors of the given
+ * node.
+ * @param num_neighbors The number of neighbors of this node.
+ *
+ * @return The pick number of the given node.
+ */
+int64_t NumPick(
+    int64_t fanout, bool replace,
+    const torch::optional<torch::Tensor>& probs_or_mask, int64_t offset,
+    int64_t num_neighbors) {
+  int64_t actual_fanout = fanout == -1 ? num_neighbors : fanout;
+  int64_t max_count =
+      probs_or_mask.has_value()
+          ? torch::count_nonzero(
+                probs_or_mask.value().slice(0, offset, offset + num_neighbors))
+                .item<int64_t>()
+          : num_neighbors;
+  if (replace) {
+    return max_count == 0 ? 0 : actual_fanout;
+  } else {
+    return std::min(max_count, actual_fanout);
+  }
```

Comment:
I noticed an inconsistency between Neighbor sample algorithm (which has been in use for long) and Labor algorithm. The number of picked neighbors is different when `fanout = -1, replace = true` and `mask` is not full zero. Under this circumstance, the Neighbor sample algorithm only picks as many as the number of neighbors with **non-zero** probability, while the picked number of Labor algorithm is the same as the number of **all** neighbors. It should be consistent. @mfbalin which way from your view is more plausible?

Reply:
Let's modify labor and equalize the behaviors. The implementation here is supposed to match neighbor fully.

Reply:
It can be formalized by saying that for `fanout=-1`, the behaviour is simply the same as passing the `num_neighbors` as the fanout for each vertex. For weighted sampling, we can say that `num_neighbors` is treated as the number of nonzero weights.

Reply:
Semantically, the fanout = -1 with replace true is non-defined. There are 2 options:
- Simply throw an exception when it happens.
- Define a consistent behavior for both sampler and document it clearly (in both c++ and python doc).

Reply:
Let's modify Labor to match Neighbor.
There is another inconsistency: when `replace = true` and `fanout` is bigger than the number of neighbors, Labor only samples `num_neighbors` neighbors. I suppose the expected result is that `fanout` neighbors are sampled with replacement.